### DISCUSSION TITLE
How to cancel tasks from SSHOperator?

### DISCUSSION THREAD

**Discussion by phwolf** [apache/airflow#17870] [Q&A] [answered]
Hi,

I need Airflow to execute some scripts on a remote machine. So far I am using the SSHOperator, which is working fine.
However, sometimes I want to cancel a task generated by the SSHOperator. For this I set the task state to failed, which kills the process on the current machine, but the process on the remote machine is still running.  I there any way to kill the process running on the remote machine using Airflow? Any suggestions / workarounds are welcome

**Comment by potiuk** [comment, marked as answer]
I think the SSH Hook might lack proper `on_kill` handling that could cleanly kill the running session. You could extend the SSHOperator to add it and then maybe even contribute it back?

**Reply by phwolf** [reply]
The SSHOperator has an argument `get_pty`, setting this to true seems to solve the problem. However in my case I had to perform further clean up tasks, so providing an `on_kill` method worked great for me.

**Reply by medina325** [reply]
Can you let me know what your code looks like? I overrided the `on_kill` method as well, but my code is not killing the remote process.
In my case I have to execute a Scrapy Spider remotely via SSHOperator, and if I set my task to failed, I need to be able to automatically kill the remote scrapy process. This is what my code looks like so far:

```python
class SpiderSSHOperator(SSHOperator):
    def on_kill(self):
        try:
            scrapy_command = self.command.split(' && ')[-1]

            pid = int(subprocess.check_output(['pgrep', '-f', scrapy_command]))
            os.kill(pid, signal)
            
            logger.info(f'Remote scrapy process with id {pid} killed')
        except subprocess.CalledProcessError as e:
            raise e
```

**Reply by jayam26** [reply]
Remote_job = SpiderSSHOperator(
                       task_id = 'Remote_job',
                       command = job_commands********,
                       ssh_hook = sshHook,
                       get_pty = True   --------------------------->  Adding this paramter dynamically  updates both in remote server and  airflow GUI
                      )

Clear and Mark failed works exactly as expected.

**Reply by medina325** [reply]
Yeah, the get_pty actually gets the job done. However my case was a bit different since Scrapy's CLI commands doesn't always return an exit code different than 0 when I want it to 😅 , which signals the operator to kill the remote process.

I'd have to wrap the command with a bash script and handle the exit codes myself, or something like that, because I never went this way.
